### PR TITLE
WIP: Discrete and synchronized targets support

### PR DESCRIPTION
**Pull request by sliwy**

### CODE REVIEW

**Comment by MFattouh** [comment]
I see a number of issues for cropped decoding with the current implementation, which uses targets as anchor (stop) points and then slices a window with the specified size before the target.

- [x] Depending on the specified window size, we could end up with windows that span a number of targets. However, currently, only the last target is returned. This is the expected behavior for Trial-wise decoding, but not in cropped decoding. We can control this behavior by adding a new optional parameter to WindowsDataset to return all targets within the window or just the last one.

- [ ] The new create_windows_from_target_channels windowing function implemented by this PR would create overlapping windows with duplicate targets, which is not desired, because targets are assumed to appear only once in a single dataset iteration. However, we can use the old windowing functions from braindecode to iterate over the entire trial, but we want to emphasize the difference between trials-wise and cropped decoding windowing functions in the documentation and the tutorial.

- [ ] Proper validation that accounts for the receptive field of the models by considering fixed number of targets (specified by the user) per trial for validation.

**Comment by robintibor** [comment]
sounds like you are thinking about the important issues, removing duplicate targets etc can always be a bit tricky, but we do have quite some logic there already in the scoring.py code, you can also take inspiration from it when you write some new code to tackle your use case

**Comment by robintibor** [comment]
So summary what I recall now from what I said.
* We can "create fixed length windows", if targets are simply inside the compute window, they are matching, we may just have to remove some at beginning if network predicts less
* If we leave NaNs inside returned targets from dataset, matching afterwards could be quite easy: creating a mask that removes NaNs on the targets and applying the same mask to the predictions should get the correct matching already
* We could make a generic loss wrapper that accepts predictions, targets and a loss function, removes the values at NaN indices as described above and then applies the loss function to the remaining values

**Comment by robintibor** [comment]
Please make sure that the naming is clear, like I just had some quick glance and the naming with ECoG seems a bit confusing to me. Keep in mind target-timeseries decoding is not directly tied to ECoG, you could do target-timeseries decoding directly from EEG as well... so make sure you don't introduce confusing naming there... Also make sure it is clear what this still has to do with BCI Competition etc., that it is clearly described.

File: braindecode/datasets/ecog_bci_competition.py [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+    References
+    ----------
+    .. [1] Schalk, G., Kubanek, J., Miller, K.J., Anderson, N.R., Leuthardt, E.C.,
+        Ojemann, J.G., Limbrick, D., Moran, D.W., Gerhardt, L.A., and Wolpaw, J.R.
+        Decoding Two Dimensional Movement Trajectories Using Electrocorticographic Signals
+        in Humans, J Neural Eng, 4: 264-275, 2007.
+    """
+    possible_subjects = [1, 2, 3]
+
+    def __init__(self, path, subject_ids=None):
```
Comment (sliwy) [inline braindecode/datasets/ecog_bci_competition.py:43]: @robintibor @agramfort For now to create this dataset user must specify path to the data (folder with .mat files downloaded from BCI IV competition site). To download the data you have to sign in with your email and password received in an email. Do you see any easier option to do that (without additional steps outside braindecode)? Do we want to create any specific path where we can put this kind of data (like mne_data)?

Comment (robintibor) [reply braindecode/datasets/ecog_bci_competition.py:43]: so this dataset is not in moabb? I mean if you have to manually download anyways I feel supplying path seems fine to me...

File: examples/plot_bcic_iv_4_ecog_trial.py [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+#
+# .. note::
+#    To load your own datasets either via mne or from
+#    preprocessed X/y numpy arrays, see `MNE Dataset
+#    Tutorial <./plot_mne_dataset_example.html>`__ and `Numpy Dataset
+#    Tutorial <./plot_custom_dataset_example.html>`__.
+#
+from sklearn.metrics import r2_score
+
+DATASET_PATH = '/home/maciej/projects/braindecode/BCICIV_4_mat'
```
Comment (robintibor) [inline examples/plot_bcic_iv_4_ecog_trial.py:38]: I assume you may have to mock the dataset like Lukas did in https://github.com/braindecode/braindecode/pull/276 for this to run on CI eventually? Or what is your plan?

File: braindecode/datasets/ecog_bci_competition.py (hunk 1 above)
Comment (sliwy) [reply braindecode/datasets/ecog_bci_competition.py:43]: This dataset is not in moabb, so probably we will keep the dataset path as a parameter

File: examples/plot_bcic_iv_4_ecog_trial.py (hunk 2 above)
Comment (sliwy) [reply examples/plot_bcic_iv_4_ecog_trial.py:38]: Yeah, I will try to prepare something similar! :)

File: braindecode/datasets/ecog_bci_competition.py [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+
+    Contains ECoG recordings for three patients moving fingers during the experiment.
+    Targets correspond to the time courses of the flexion of each of five fingers.
+    See http://www.bbci.de/competition/iv/desc_4.pdf for the dataset description.
+
+    Test labels can be downloaded from:
+    http://www.bbci.de/competition/iv/results/ds4/true_labels.zip
+
+    Data can be downloaded from (Dataset 4):
+    http://www.bbci.de/competition/iv/#dataset4
```
Comment (agramfort) [inline braindecode/datasets/ecog_bci_competition.py:19]: these data are super old. I am sure we could now make them available without password eg storing them on OSF.

I would contact the authors of the competition to ask for permission.

Maybe @cbrnr nows who to ask this.

I really don't like that an example does not work out of the box. And you won't be able to test it in CI or to build any doc. It's for me a blocker.

**Comment by agramfort** [comment]
let's talk about the design of this one on discord when you have time.

File: braindecode/datasets/ecog_bci_competition.py (hunk 3 above)
Comment (cbrnr) [reply braindecode/datasets/ecog_bci_competition.py:19]: Not sure who "owns" the data, but if it's BBCI I'd contact Benjamin Blankertz. If not, either Gerwin Schalk or Kai J. Miller are listed. I know Benjamin and Gerwin, if you want I could ask them.

**Comment by agramfort** [comment]
can you send an email to Benjamin and Gerwin with us in CC to explain what
we do here?

thx !

>

**Comment by cbrnr** [comment]
Sure, will do!

**Comment by agramfort** [comment]
@sliwy the fetcher should pull the data from https://searchworks.stanford.edu/view/zk881ps0522
no registration required.

**Comment by sliwy** [comment]
@agramfort I've seen the email, it's perfect. For the implementation, should we create `braindecode_data` folder to save the data like in mne or can we use `mne_data` folder like moabb even if it's not meant for data from mne? 

@cbrnr Thank you for  your help! :)

**Comment by robintibor** [comment]
Please let me also know if you need any support/have any questions at the moment with this @MFattouh @sliwy :)

**Comment by sliwy** [comment]
@robintibor that's a good question. Before the weekend I started working on loading the dataset from the repository mentioned above. Currently I'm on vacations and I didn't have enough time to work together with @MFattouh. Next week should be a little bit easier than last weeks, so I hope we can make it. :)

**Comment by sliwy** [comment]
A quick update on the PR, I asked Gerwin Schalk about another possible solution to have this dataset openly available. From the link provided by them we can download only a bigger dataset that was probably used to prepare the data. I f we want to use that data repository it would need from us some additional steps of data extraction, downloading additional not needed data and more information from dataset providers.

Instead I asked Gerwin Schalk if we can just take the files from BCI Competition and put them on OSF (as @agramfort suggested). We will give credits to the owners of the dataset and with much smaller effort we will have the dataset openly available. Still there are some code changes to be done in the meantime finish the PR.

**Comment by robintibor** [comment]
so how it's going any support needed @sliwy @MFattouh ?

**Comment by sliwy** [comment]
Great that you ask, so for now a few things are clear:
1. I had some discussions with Kai J Miller about the dataset and its availability. Two days ago the dataset was uploaded to the ECoG library, so we finally have an easy access to the dataset. I have already modified the dataset downloading part to use new ECoG library instead of path to earlier downloaded files. This should be fine.
2. I think that trialwise example is almost ready. I added some tests yesterday, I'll take one more look at it (to add some more tests) this week.
3. For the cropped decoding things are more complicated but still quite close to the end. I think that we need some improvement in the last part of the tutorial where there are some big code snippets with for-loops etc. It can be placed inside functions but I feel that @MFattouh can have more to explain that. Of course @robintibor you can take a look at cropped decoding example as you're the master of cropped decoding. :)
4. @robintibor do you know why there is no documentation built based on this PR? I didn't have time to check that but I see that there is only one check done here.

File: braindecode/datasets/ecog_bci_competition.py (hunk 1 above)
Comment (sliwy) [reply braindecode/datasets/ecog_bci_competition.py]: This dataset will be stored in `mne_data` as MOABB datasets, so problem solved :)

File: examples/plot_bcic_iv_4_ecog_trial.py (hunk 2 above)
Comment (sliwy) [reply examples/plot_bcic_iv_4_ecog_trial.py]: This one is still to be done. Mocking and testing BCICompetitionDataset4

**Comment by robintibor** [comment]
Maybe the whole draft PR logic prevents CI from running? I have remove this draft stuff. Maybe still a commit needed  to make it run? Or needs rebase...

**Comment by robintibor** [comment]
Well might anyways be a good step to try to rebase @sliwy ? Then things may run?

File: examples/plot_bcic_iv_4_ecog_cropped.py [hunk 4]
```diff
@@ -0,0 +1,358 @@
...
+######################################################################
+# Loading
+# ~~~~~~~
+#
+
+
+######################################################################
+# First, we load the data. In this tutorial, we use the functionality of
+# braindecode to load datasets through
+# `MOABB <https://github.com/NeuroTechX/moabb>`__ to load the BCI
```
Comment (robintibor) [inline examples/plot_bcic_iv_4_ecog_cropped.py:34]: This text and also a lot of other text in this example is not correct/still fits to old example. Feel free to also shorten text quite a bit, but at least the text that is there should be correct.

**Comment by robintibor** [comment]
I made some small comments in the cropped example @sliwy. Also for final for loops, could https://github.com/braindecode/braindecode/blob/2dbd40cc7951934ad32892770536e360b6188920/braindecode/training/scoring.py#L303 help there? Otherwise what is the status, are there things still open that you want to do? @sliwy @MFattouh

**Comment by sliwy** [comment]
@robintibor this `predict_trials` function may be really helpful to make cropped decoding example more accessible. Refactor of the cropped decoding example still on my list. :)

**Comment by robintibor** [comment]
@sliwy feel free to adapt the predict_trials function to be more flexible if that is causing the error

File: examples/plot_bcic_iv_4_ecog_cropped.py [hunk 5]
```diff
@@ -0,0 +1,381 @@
...
+from sklearn.model_selection import train_test_split
+
+# We can split train dataset into training and validation datasets using
+# `sklearn.model_selection.train_test_split` and `torch.utils.data.Subset`
+idx_train, idx_valid = train_test_split(np.arange(len(train_set)),
+                                        random_state=100,
+                                        test_size=0.2,
+                                        shuffle=False)
+valid_set = test_set
+# valid_set = torch.utils.data.Subset(train_set, idx_valid)
```
Comment (sliwy) [inline examples/plot_bcic_iv_4_ecog_cropped.py:236]: @robintibor is there any way to subset a cropped dataset without raising error while validating that a windows indices have to start from zero?

Comment (robintibor) [reply examples/plot_bcic_iv_4_ecog_cropped.py:236]: hmhm very good question. I guess no and we maybe should implement it? I think usually our expectation was that splitting is mostly done before windowing. However, I have also split valid datasets off windows/cropped datasets, but there usually I knew how many windows there were per trial and could ensure by hand that everything will be correct.. probably we should have some function for it? But to not delay this PR further is it possible to move the splitting to before the windowing?

File: examples/plot_bcic_iv_4_ecog_cropped.py [hunk 6]
```diff
@@ -0,0 +1,381 @@
...
+    )
+print('Correlation coefficient for each dimension: ', corr_coeffs)
+
+# We plot predicted and target finger flexion
+x_time = np.linspace(0, preds_test.shape[1] / sfreq, preds_test.shape[1])
+fig, ax1 = plt.subplots(figsize=(8, 3))
+ax1.plot(x_time, y_test[0, :], label='Target')
+ax1.plot(x_time, preds_test[0, :], label='Predicted')
+ax1.xlabel('Time [s]')
+ax1.ylabel('Finger flexion')
```
Comment (robintibor) [inline examples/plot_bcic_iv_4_ecog_cropped.py:380]:
```suggestion
ax1.set_xlabel('Time [s]')
ax1.set_ylabel('Finger flexion')
```

needs to be corrected @sliwy

**Comment by sliwy** [comment]
@robintibor Two things that left to be done:
1. Improve tutorial appearance (sections, spacing etc.)
2. Change splitting in the cropped decoding (split on raw datasets).

Probably this weekend will be enough to finish. :)

**Comment by robintibor** [comment]
Great would be amazing @sliwy ! Then we can move towards next release

File: braindecode/datasets/bcicomp.py [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+        will be automatically downloaded to the specified folder.
+        force_update (bool) – Force update of the dataset even if a local copy exists.
+        verbose (bool, str, int, or None) – If not None, override default verbose level
+        (see mne.verbose())
+
+        Returns
+        -------
+
+        """
+        sign = 'BCICompetitionIVDataset4'
```
Comment (robintibor) [inline braindecode/datasets/bcicomp.py:93]: sign?

File: examples/plot_bcic_iv_4_ecog_trial.py [hunk 8]
```diff
@@ -0,0 +1,389 @@
...
+    model,
+    criterion=torch.nn.MSELoss,
+    optimizer=torch.optim.AdamW,
+    train_split=predefined_split(valid_set),  # using valid_set for validation,
+    optimizer__lr=lr,
+    optimizer__weight_decay=weight_decay,
+    batch_size=batch_size,
+    callbacks=[
+        'r2',
+        ('valid_pearson_r', EpochScoring(pearson_r_score, lower_is_better=False,
```
Comment (robintibor) [inline examples/plot_bcic_iv_4_ecog_trial.py:282]:
```suggestion
        ('valid_pearson_r', EpochScoring(pearson_r_score, lower_is_better=False, on_train=False,
```

more symmetric

**Comment by robintibor** [comment]
so @sliwy how is going? so this is ready to merge for you? did you see small comments from me?

**Comment by sliwy** [comment]
@robintibor I was a little bit busy, from my side it's ready to merge, I fixed the comments from you, thanks :)

File: braindecode/training/scoring.py [hunk 9]
```diff
@@ -206,14 +207,96 @@ def on_epoch_end(self, net, dataset_train, dataset_valid, **kwargs):
...
+            # timeseries of predictions/targets of shape (n_classes x timesteps)
+
+            # mask NaNs form targets
+            preds = np.hstack(trial_preds)  # n_classes x timesteps in all trials
+            targets = np.hstack(trial_ys)
+            # create valid targets mask
+            mask = ~np.isnan(targets)
+            # select valid targets that have a matching predictions
+            masked_targets = targets[mask]
+            if mask.shape[0] == 1:
```
Comment (robintibor) [inline braindecode/training/scoring.py:272]: could you put a comment here when/why this case happens? @sliwy

**Comment by robintibor** [comment]
Great work @sliwy! I had a look through now, could you add the comment, in that position I was not sure why which case appears, and would be great to have it documented there :) 
After, I would merge. then we could do  a new braindecode release soon after ICLR deadline next tuesday!

In the future we may wanna think about adding a 'target_channels' argument  to the `create_windows_from_target_channels` to ease usability using that API (keeping WindowsDataset same with purely `targets_from` logic). But for now seems all fine to me!

File: braindecode/datasets/bcicomp.py [hunk 10]
```diff
@@ -91,18 +93,18 @@ def download(path=None, force_update=False, verbose=None):
...
         signature = 'BCICompetitionIVDataset4'
+        folder_name = 'BCI_Competion4_dataset4_data_fingerflexions'
         # Check if the dataset already exists (unpacked). We have to do that manually
         # because we are removing .zip file from disk to save disk space.
         path = get_dataset_path(signature, path)
         key_dest = "MNE-{:s}-data".format(signature.lower())
-        destination = _url_to_local_path(DATASET_URL, osp.join(path, key_dest))[
-                      :-4] + '/BCI_Competion4_dataset4_data_fingerflexions'
-        if len(list(glob.glob(destination + '/*.mat'))) == 6:
+        # We do not use mne _url_to_local_path due to ':' in the url that causes problems on Windows
```
Comment (robintibor) [inline braindecode/datasets/bcicomp.py:101]: interesting, maybe also interesting to know for @agramfort ?

**Comment by robintibor** [comment]
Amazing amazing @sliwy merged
